tests/zfs-tests/tests/functional/dedup/dedup_quota.ksh

```diff
@@ -0,0 +1,221 @@
+#!/bin/ksh -p
+# CDDL HEADER START
+#
+# The contents of this file are subject to the terms of the
+# Common Development and Distribution License (the "License").
+# You may not use this file except in compliance with the License.
+#
+# You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+# or https://opensource.org/licenses/CDDL-1.0.
+# See the License for the specific language governing permissions
+# and limitations under the License.
+#
+# When distributing Covered Code, include this CDDL HEADER in each
+# file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+# If applicable, add the following below this CDDL HEADER, with the
+# fields enclosed by brackets "[]" replaced with your own identifying
+# information: Portions Copyright [yyyy] [name of copyright owner]
+#
+# CDDL HEADER END
+#
+
+#
+# Copyright (c) 2023, Klara Inc.
+#
+
+# DESCRIPTION:
+#	Verify that new entries are not added to the DDT when dedup_table_quota has
+#	been exceeded.
+#
+# STRATEGY:
+#	1. Create a pool with dedup=on
+#	2. Set threshold for on-disk DDT via dedup_table_quota
+#	3. Verify the threshold is exceeded after zpool sync
+#	4. Verify no new entries are added after subsequent sync's
+#	5. Remove all but one entry from DDT
+#	6. Verify new entries are added to DDT
+#
+
+. $STF_SUITE/include/libtest.shlib
+. $STF_SUITE/tests/functional/events/events_common.kshlib
+
+verify_runnable "both"
+
+log_assert "DDT quota is enforced"
+
+MOUNTDIR="$TEST_BASE_DIR/dedup_mount"
+FILEPATH="$MOUNTDIR/dedup_file"
+VDEV_GENERAL="$TEST_BASE_DIR/vdevfile.general.$$"
+VDEV_DEDUP="$TEST_BASE_DIR/vdevfile.dedup.$$"
+POOL="dedup_pool"
+
+ORIG_TXG_TIMEOUT=$(get_tunable TXG_TIMEOUT)
+
+log_note ddt_zap_default_bs = $ORIG_DDT_ZAP_DEFAULT_BS
+
+function cleanup
+{
+	if poolexists $POOL ; then
+		destroy_pool $POOL
+	fi
+	log_must rm -fd $VDEV_GENERAL $VDEV_DEDUP $MOUNTDIR
+	log_must set_tunable32 TXG_TIMEOUT $ORIG_TXG_TIMEOUT
+}
+
+
+function do_clean
+{
+	log_must zpool destroy $POOL
+	log_must rm -fd $VDEV_GENERAL $VDEV_DEDUP $MOUNTDIR
+}
+
+function do_setup
+{
+	# Create a general VDEV and special dedup VDEV
+	log_must truncate -s 5G $VDEV_GENERAL
+	log_must truncate -s 200M $VDEV_DEDUP
```

Comment:
It seems odd to create `VDEV_DEDUP` here instead of where it's added to the pool.

Reply:
moved to pool add section